transaction/src/action/output.rs

```diff
@@ -22,6 +25,45 @@ impl IsAction for Output {
     fn balance_commitment(&self) -> balance::Commitment {
         self.body.balance_commitment
     }
+
+    fn decrypt_with_perspective(
+        &self,
+        txp: &TransactionPerspective,
+    ) -> anyhow::Result<Option<ActionView>> {
+        // Get payload key for note commitment of note payload
+
+        let note_commitment = self.body.note_payload.note_commitment;
+
+        // Get payload key for note commitment of swap NFT.
+        let payload_key = txp
+            .payload_keys
+            .get(&note_commitment)
+            .ok_or_else(|| anyhow::anyhow!("corresponding payload key not found"))?;
+
+        // Decrypt note
+
+        let decrypted_note =
+            Note::decrypt_with_payload_key(&self.body.note_payload.encrypted_note, payload_key)?;
+        // If memo has not been decrypted yet
+        // * Decrypt wrapped_memo_key
+
+        let decrypted_memo_key = self.body.wrapped_memo_key.decrypt_outgoing(payload_key)?;
+
+        // * Decrypt memo using wrapped memo key
+
+        let memo_cipher_text = txp
+            .memo_cipher_text
+            .as_ref()
+            .ok_or_else(|| anyhow::anyhow!("no memo present"))?
+            .to_owned();
+
+        let memo = MemoPlaintext::decrypt(memo_cipher_text, &decrypted_memo_key)?;
+
+        Ok(Some(ActionView::Output(OutputView {
+            decrypted_note,
+            memo,
+        })))
+    }
```

Comment:
This method is never going to return `Ok(None)`, right?  Doesn't that mean that we'll error if a transaction has an output that's not viewable from our `Perspective`?

Reply:
I suppose so - signature can change accordingly.